// @roots/bud-server/lib/cjs/server/server.https.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Https = void 0;
const bud_support_1 = require("@roots/bud-support");
const https_1 = require("https");
const server_base_1 = require("./server.base");
const { readFile } = bud_support_1.fs;
/**
 * HTTP Server
 *
 * @public
 */
class Https extends server_base_1.BaseServer {
    constructor() {
        super(...arguments);
        /**
         * Create HTTPS server
         *
         * @public
         */
        this.createServer = async function (app) {
            const key = await this.getKey();
            const cert = await this.getCert();
            this.instance = (0, https_1.createServer)({ key, cert }, app);
            return this.instance;
        };
    }
    /**
     * Has SSL key
     *
     * @public
     */
    hasKey() {
        return this.app.hooks.filter('dev.ssl.key') ? true : false;
    }
    /**
     * Get SSL key
     *
     * @returns
     */
    async getKey() {
        !this.hasKey() && this.app.error('Server key is not defined');
        return await readFile(this.app.hooks.filter('dev.ssl.key'), 'utf8');
    }
    /**
     * Has SSL certificate
     *
     * @public
     */
    hasCert() {
        return this.app.hooks.filter('dev.ssl.cert') ? true : false;
    }
    /**
     * Get SSL certificate
     *
     * @public
     */
    async getCert() {
        !this.hasCert() && this.app.error('Server cert is not defined');
        return await readFile(this.app.hooks.filter('dev.ssl.cert'), 'utf8');
    }
}
exports.Https = Https;
