// @roots/bud-framework/lib/cjs/Logger/logger.constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.types = exports.LEVEL = exports.INSTANCE_CONFIG = void 0;
const figures_1 = require("./figures");
/**
 * Instance configuration
 *
 * @internal
 */
exports.INSTANCE_CONFIG = {
    displayScope: true,
    displayBadge: true,
    displayDate: false,
    displayFilename: false,
    displayLabel: false,
    displayTimestamp: false,
    underlineLabel: false,
    underlineMessage: false,
    underlinePrefix: false,
    underlineSuffix: false,
    uppercaseLabel: false,
};
exports.LEVEL = {
    vvvv: 'log',
    vvv: 'timer',
    vv: 'warn',
    v: 'error',
};
const types = () => ({
    error: {
        badge: figures_1.figures.cross,
        color: 'red',
        label: 'error',
        logLevel: exports.LEVEL['v'],
    },
    fatal: {
        badge: figures_1.figures.cross,
        color: 'red',
        label: 'fatal',
        logLevel: exports.LEVEL['v'],
    },
    star: {
        badge: figures_1.figures.star,
        color: 'cyan',
        label: 'instantiate',
        logLevel: exports.LEVEL['vvv'],
    },
    info: {
        badge: figures_1.figures.info,
        color: 'magenta',
        label: 'log',
        logLevel: exports.LEVEL['vvv'],
    },
    success: {
        badge: figures_1.figures.tick,
        color: 'green',
        label: 'success',
        logLevel: exports.LEVEL['vvv'],
    },
    warn: {
        badge: figures_1.figures.warning,
        color: 'yellow',
        label: 'warning',
        logLevel: exports.LEVEL['vv'],
    },
    complete: {
        badge: figures_1.figures.circleFilled,
        color: 'cyan',
        label: 'complete',
        logLevel: exports.LEVEL['vvv'],
    },
    pending: {
        badge: figures_1.figures.ellipsis,
        color: 'cyan',
        label: 'pending',
        logLevel: exports.LEVEL['vvv'],
        stream: process.stdout,
    },
    note: {
        badge: figures_1.figures.bullet,
        color: 'blue',
        label: 'note',
        logLevel: exports.LEVEL['vvv'],
    },
    start: {
        badge: figures_1.figures.play,
        color: 'green',
        label: 'start',
        logLevel: exports.LEVEL['vvv'],
    },
    pause: {
        badge: figures_1.figures.squareSmallFilled,
        color: 'yellow',
        label: 'pause',
        logLevel: exports.LEVEL['vvv'],
    },
    debug: {
        badge: figures_1.figures.circleFilled,
        color: 'red',
        label: 'log',
        logLevel: exports.LEVEL['vvvv'],
    },
    await: {
        badge: figures_1.figures.ellipsis,
        color: 'cyan',
        label: 'awaiting',
        logLevel: exports.LEVEL['vvv'],
    },
    watch: {
        badge: figures_1.figures.ellipsis,
        color: 'yellow',
        label: 'watching',
        logLevel: exports.LEVEL['vvv'],
    },
    log: {
        badge: figures_1.figures.pointer,
        color: 'blue',
        label: 'log',
        logLevel: exports.LEVEL['vvv'],
    },
});
exports.types = types;
