// @roots/bud/lib/cjs/factory/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.factory = void 0;
const Bud_1 = require("../Bud");
const seed_1 = require("../seed");
const services_1 = require("../services");
/**
 * Create a {@link Bud} instance programatically
 *
 * @example
 * ```ts
 * const bud = factory()
 * ```
 *
 * @returns Bud instance
 *
 * @public
 */
async function factory(overrides) {
    const options = {
        ...(overrides ?? {}),
        services: {
            ...services_1.services,
            ...(overrides?.services ?? {}),
        },
        config: {
            ...seed_1.seed,
            ...(overrides?.config ?? {}),
            features: {
                ...seed_1.seed.features,
                ...(overrides?.config?.features ?? {}),
            },
            location: {
                ...seed_1.seed.location,
                ...(overrides?.config?.location ?? {}),
            },
            build: {
                ...seed_1.seed.build,
                ...(overrides?.config?.build ?? {}),
            },
        },
    };
    process.env.BABEL_ENV = options.config.mode;
    process.env.NODE_ENV = options.config.mode;
    const project = new Bud_1.Bud(options);
    project.time(project.name);
    project.log({
        message: 'process.env.NODE_ENV',
        suffix: process.env.NODE_ENV,
    });
    project.log({
        message: 'process.env.BABEL_ENV',
        suffix: process.env.NODE_ENV,
    });
    await project.lifecycle();
    return project;
}
exports.factory = factory;
