// @roots/bud-compiler/lib/cjs/Compiler/process/stats.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.write = exports.theme = void 0;
const bud_support_1 = require("@roots/bud-support");
exports.theme = {
    foregroundColor: '#eff0eb',
    backgroundColor: '#282a36',
    red: '#ff5c57',
    green: '#5af78e',
    yellow: '#f3f99d',
    blue: '#57c7ff',
    magenta: '#ff6ac1',
    cyan: '#9aedfe',
};
function makeTable(data) {
    return bud_support_1.table.table(data, {
        border: bud_support_1.table.getBorderCharacters('void'),
        columnDefault: {
            alignment: 'left',
            paddingLeft: 2,
            paddingRight: 2,
        },
        columns: [
            { alignment: 'left' },
            { alignment: 'left' },
            { alignment: 'left' },
        ],
    });
}
function write(stats) {
    const compilers = stats.children?.map(compilation => {
        if (!compilation?.entrypoints)
            return compilation;
        const errors = compilation.errors?.map(error => {
            return (0, bud_support_1.boxen)(`\n${error.message}`, {
                title: `${error.title ?? 'error'}`,
                margin: {
                    top: 0,
                    bottom: 1,
                    left: 0,
                    right: 0,
                },
                padding: {
                    top: 0,
                    bottom: 0,
                    right: 0,
                    left: 0,
                },
                borderColor: exports.theme.red,
            });
        });
        const warnings = compilation.warnings?.map(warning => {
            return (0, bud_support_1.boxen)(`\n${warning.message}`, {
                title: `${warning.title ?? 'warning'}`,
                margin: {
                    top: 0,
                    bottom: 1,
                    left: 0,
                    right: 0,
                },
                padding: {
                    top: 0,
                    bottom: 0,
                    right: 0,
                    left: 0,
                },
                borderColor: exports.theme.yellow,
            });
        });
        const assets = (0, bud_support_1.boxen)(makeTable([
            [' name', 'hot', 'size'].map(i => bud_support_1.chalk.bold.hex(exports.theme.magenta)(i)),
            ...compilation.assets
                ?.filter(({ emitted }) => emitted)
                .map(asset => [
                bud_support_1.chalk.hex(asset.info.error
                    ? exports.theme.red
                    : asset.info.warn
                        ? exports.theme.yellow
                        : '#FFFFFF')(` ${asset.name}`),
                asset.info.hotModuleReplacement
                    ? bud_support_1.chalk.hex(exports.theme.green)('✔')
                    : bud_support_1.chalk.dim('✘'),
                bud_support_1.humanReadable.sizeFormatter()(asset.info.size),
            ]),
        ]), {
            title: `assets`,
            margin: {
                top: 1,
                bottom: 1,
                left: 0,
                right: 0,
            },
            padding: {
                left: 0,
                top: 1,
                right: 0,
                bottom: 0,
            },
            borderColor: exports.theme.blue,
        });
        return {
            ...compilation,
            boxes: [
                compilation.errorsCount ? errors.join('\n') : null,
                compilation.warningsCount ? warnings.join('\n') : null,
                compilation.assets.filter(({ emitted }) => emitted).length
                    ? assets
                    : null,
            ].filter(Boolean),
        };
    });
    const out = compilers?.map(compiler => compiler.boxes.join(''));
    // eslint-disable-next-line no-console
    console.log(out.join(''));
}
exports.write = write;
