// @roots/bud-server/lib/cjs/client/overlay/inner.web-component.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Inner = void 0;
class Inner extends HTMLElement {
    constructor() {
        super(...arguments);
        /**
         * Inner HTML
         *
         * @public
         */
        this.innerHTML = ``;
    }
    /**
     * Render component
     *
     * @public
     */
    render() {
        this.innerHTML = `
      ${this.innerHTML}
    `;
        this.style.position = `relative`;
        this.style.display = `flex`;
        this.style.flexWrap = `flex-wrap`;
        this.style.flexDirection = `column`;
        this.style.alignContent = `center`;
        this.style.justifyContent = `center`;
        this.style.alignItems = `center`;
        this.style.justifyItems = `center`;
        this.style.padding = `0.5rem`;
        this.style.margin = `1rem`;
        this.style.height = `100%`;
        this.style.width = `100%`;
        this.style.maxWidth = `100%`;
        this.style.maxHeight = `100%`;
    }
    /**
     * Connected callback
     *
     * @public
     */
    connectedCallback() {
        if (!this.rendered) {
            this.render();
            this.rendered = true;
        }
    }
    /**
     * Attribute changed callback
     *
     * @public
     */
    attributeChangedCallback() {
        this.render();
    }
}
exports.Inner = Inner;
