// @roots/bud-server/lib/cjs/client/indicator/indicator.component.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.IndicatorComponent = void 0;
const pulse_component_1 = require("./pulse.component");
/**
 * Indicator web component
 *
 * @public
 */
class IndicatorComponent extends HTMLElement {
    constructor() {
        super(...arguments);
        /**
         * Component name
         *
         * @public
         */
        this.name = `bud-activity-indicator`;
        /**
         * Status indicator colors
         *
         * @public
         */
        this.colors = {
            success: [4, 120, 87],
            error: [220, 38, 38],
            warn: [252, 211, 77],
            pending: [59, 130, 246],
        };
    }
    /**
     * Get accessor: has errors
     *
     * @public
     */
    get hasErrors() {
        return this.getAttribute('has-errors') == 'true';
    }
    /**
     * Get accessor: has warnings
     *
     * @public
     */
    get hasWarnings() {
        return this.getAttribute('has-warnings') == 'true';
    }
    /**
     * Compilation is ongoing
     *
     * @public
     */
    get isPending() {
        return (!this.payload?.errors?.length &&
            !this.payload?.warnings?.length &&
            this.payload?.action == 'building');
    }
    /**
     * Render status indicator
     *
     * @public
     */
    render() {
        this.classList.add(this.name);
        this.innerHTML = `
    <style>
      .${this.name} {
        position: fixed;
        width: 10px;
        height: 10px;
        left: 10px;
        bottom: 10px;
        z-index: 9998;
        margin: 10px;
        padding: 5px;
        transition: opacity ease 1500ms;
        pointer-events: none;
        border-radius: 50%;
      }

      ${(0, pulse_component_1.pulse)(`${this.name}__success`, this.colors.success)}
      ${(0, pulse_component_1.pulse)(`${this.name}__error`, this.colors.error)}
      ${(0, pulse_component_1.pulse)(`${this.name}__warning`, this.colors.warn)}
      ${(0, pulse_component_1.pulse)(`${this.name}__pending`, this.colors.pending)}

      .${this.name}__visible {
        opacity: 1;
      }

      .${this.name}__hidden {
        opacity: 0;
      }
    </style>
    `;
    }
    /**
     * Show status indicator
     *
     * @public
     */
    show() {
        clearTimeout(this.hideTimeout);
        this.classList.remove(`${this.name}__hidden`);
    }
    /**
     * Hide status indicator
     */
    hide() {
        this.hideTimeout = setTimeout(() => {
            this.classList.remove(`${this.name}__error`, `${this.name}__warning`, `${this.name}__success`, `${this.name}__pending`);
            this.classList.add(`${this.name}__hidden`);
        }, 2000);
    }
    /**
     * Status is pending
     *
     * @public
     */
    pending() {
        this.show();
        this.classList.remove(`${this.name}__error`, `${this.name}__warning`, `${this.name}__success`);
        this.classList.add(`${this.name}__pending`);
        this.hide();
    }
    /**
     * Status is success
     *
     * @public
     */
    success() {
        this.show();
        this.classList.remove(`${this.name}__error`, `${this.name}__warning`, `${this.name}__pending`);
        this.classList.add(`${this.name}__success`);
        this.hide();
    }
    /**
     * Status is error
     *
     * @public
     */
    error() {
        this.show();
        this.classList.remove(`${this.name}__warning`, `${this.name}__success`, `${this.name}__pending`);
        this.classList.add(`${this.name}__error`);
    }
    /**
     * Status is warning
     *
     * @public
     */
    warning() {
        this.show();
        this.classList.remove(`${this.name}__error`, `${this.name}__success`, `${this.name}__pending`);
        this.classList.add(`${this.name}__warning`);
    }
    /**
     * Update status
     *
     * @public
     */
    update() {
        if (!this.payload?.errors?.length &&
            !this.payload?.warnings?.length &&
            this.payload.action == 'built') {
            this.success();
            return;
        }
        if (this.payload?.action == 'building' ||
            this.payload?.action == 'sync') {
            this.pending();
            return;
        }
        if (this.payload?.errors?.length) {
            this.error();
            return;
        }
        if (this.payload?.warnings?.length) {
            this.warning();
            return;
        }
    }
    static get observedAttributes() {
        return ['has-errors', 'has-warnings', 'action'];
    }
    attributeChangedCallback() {
        this.update();
    }
    connectedCallback() {
        if (!this.rendered) {
            this.render();
            this.rendered = true;
        }
    }
}
exports.IndicatorComponent = IndicatorComponent;
