// @roots/bud-build/lib/cjs/Build/loaders.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const mini_css_extract_plugin_1 = __importDefault(require("mini-css-extract-plugin"));
const Loader_1 = require("../Loader");
exports.default = {
    /**
     * Returns {@link Loader} from `css-loader`
     * @public
     */
    css: () => new Loader_1.Loader(require.resolve('css-loader')),
    /**
     * Returns {@link Loader} for `csv-loader`
     * @public
     */
    csv: () => new Loader_1.Loader(require.resolve('csv-loader')),
    /**
     * Returns {@link Loader} for `file-loader`
     * @public
     */
    file: () => new Loader_1.Loader(require.resolve('file-loader')),
    /**
     * Returns {@link Loader} for `html-loader`
     * @public
     */
    html: () => new Loader_1.Loader(require.resolve('html-loader')),
    /**
     * Returns {@link Loader} for `remark-loader`
     * @public
     */
    md: () => new Loader_1.Loader(require.resolve('remark-loader')),
    /**
     * Returns {@link Loader} for `mini-css-extract-plugin.loader`
     * @public
     */
    minicss: () => new Loader_1.Loader(mini_css_extract_plugin_1.default.loader),
    /**
     * Returns {@link Loader} for `resolve-url-loader`
     * @public
     */
    ['resolve-url']: () => new Loader_1.Loader(require.resolve('resolve-url-loader')),
    /**
     * Returns {@link Loader} for `style-loader`
     * @public
     */
    style: () => new Loader_1.Loader(require.resolve('style-loader')),
    /**
     * Returns {@link Loader} for `url-loader`
     * @public
     */
    url: () => new Loader_1.Loader(require.resolve('url-loader')),
    /**
     * Returns {@link Loader} for `xml-loader`
     * @public
     */
    xml: () => new Loader_1.Loader(require.resolve('xml-loader')),
};
