// @roots/bud-babel/lib/cjs/babel.extension.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.register = exports.mixin = exports.name = void 0;
const bud_build_1 = require("@roots/bud-build");
const babel_config_1 = require("./babel.config");
const babel_constants_1 = require("./babel.constants");
/**
 * Adds Babel transpiler support to Framework projects
 *
 * @public
 */
exports.name = '@roots/bud-babel';
/**
 * Exposes app.babel configuration utility
 *
 * @public
 */
const mixin = async (app) => ({
    babel: [babel_config_1.Config, app],
});
exports.mixin = mixin;
/**
 * Extension register event
 *
 * @public
 */
const register = async (app) => {
    app.build.loaders.babel = new bud_build_1.Loader(() => require.resolve('babel-loader'));
    app.build.items.babel = new bud_build_1.Item({
        loader: ({ build }) => build.loaders.babel,
        options: app => {
            const options = {
                cacheDirectory: app.path('storage', 'cache', 'babel'),
                env: {
                    development: {
                        compact: false,
                    },
                },
                root: app.path('src'),
            };
            if (app.babel?.presets) {
                options.presets = Object.values(app.babel.presets);
            }
            if (app.babel?.plugins) {
                options.plugins = Object.values(app.babel.plugins);
            }
            return options;
        },
    });
    app.build.rules.js.setUse([app.build.items.babel]);
    app.babel.setPresets(babel_constants_1.DEFAULT_PRESETS).setPlugins(babel_constants_1.DEFAULT_PLUGINS);
};
exports.register = register;
